### PR TITLE
[java] Adds support for fp16 and bf16 tensors

### LINKED ISSUE #7003: create tensor of float16 as input using the Java API
Hi team,

In order to run a fp16 model, do we have a way to create a tensor of float16 as input using the Java API? e.g., 
OnnxTensor.createTensor(env, FloatBuffer.wrap(pixels), shape, ***);

Thanks,

### PR DESCRIPTION
**Pull request by Craigacp**
### Description
The Java API currently only supports fp16 output tensors which it automatically casts to floats on the way out. This PR adds support for creating fp16 and bf16 tensors (from `java.nio.Buffer` objects or as the output of models, creation from Java short arrays is not supported), along with efficient methods for casting `FloatBuffer` into `ShortBuffer` filled with fp16 or bf16 values and vice versa.

The fp16 conversions use a trick to pull in the efficient conversion methods added to Java 20, falling back to ports of the MLAS methods otherwise. The Java 20 methods can be special cased by the C2 JIT compiler to emit the single instruction on x86 and ARM which converts fp32<->fp16, or the vectorized versions thereof, so they should be quite a bit faster than the MLAS ported one.

### Motivation and Context
fp16 and bf16 are increasingly popular formats and we've had several requests for this functionality. Fixes #7003.

cc @yuslepukhin  @cassiebreviu

### CODE REVIEW

File: java/src/main/native/ai_onnxruntime_OnnxTensor.c [hunk 1]
```diff
@@ -311,7 +304,7 @@ JNIEXPORT jshort JNICALL Java_ai_onnxruntime_OnnxTensor_getShort
     (void) jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
     const OrtApi* api = (const OrtApi*) apiHandle;
   ONNXTensorElementDataType onnxType = convertToONNXDataFormat(onnxTypeInt);
-  if ((onnxType == ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT16) || (onnxType == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT16)) {
+  if ((onnxType == ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT16) || (onnxType == ONNX_TENSOR_ELEMENT_DATA_TYPE_INT16) || (onnxType == ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16) || (onnxType == ONNX_TENSOR_ELEMENT_DATA_TYPE_BFLOAT16)) {
```
Comment (skottmckay) [inline java/src/main/native/ai_onnxruntime_OnnxTensor.c:307]: nit: please keep lines to 120 chars max

File: java/src/test/java/ai/onnxruntime/ModelGenerators.java [hunk 2]
```diff
@@ -181,4 +181,74 @@ public void generateMatMul() throws IOException {
...
+    matmul.setOpType("Cast");
+    matmul.addInput("input");
+    matmul.addOutput("output");
+    matmul.addAttribute(
+        OnnxMl.AttributeProto.newBuilder()
+            .setName("to")
+            .setType(OnnxMl.AttributeProto.AttributeType.INT)
+            .setI(outputDataType.getNumber())
+            .build());
+    graph.addNode(matmul);
```
Comment (skottmckay) [inline java/src/test/java/ai/onnxruntime/ModelGenerators.java:219]:
```suggestion
    OnnxMl.NodeProto.Builder cast = OnnxMl.NodeProto.newBuilder();
    cast.setName("cast-0");
    cast.setOpType("Cast");
    cast.addInput("input");
    cast.addOutput("output");
    cast.addAttribute(
        OnnxMl.AttributeProto.newBuilder()
            .setName("to")
            .setType(OnnxMl.AttributeProto.AttributeType.INT)
            .setI(outputDataType.getNumber())
            .build());
    graph.addNode(cast);
```

**Review by skottmckay — DISMISSED** [review]

File: java/src/main/java/ai/onnxruntime/OnnxJavaType.java [hunk 3]
```diff
@@ -17,9 +17,11 @@ public enum OnnxJavaType {
   BOOL(7, boolean.class, 1),
   STRING(8, String.class, 4),
   UINT8(9, byte.class, 1),
+  FLOAT16(10, short.class, 2),
+  BFLOAT16(11, short.class, 2),
   UNKNOWN(0, Object.class, 0);
 
-  private static final OnnxJavaType[] values = new OnnxJavaType[10];
+  private static final OnnxJavaType[] values = new OnnxJavaType[13];
```
[the remark this answers is not in the document]
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxJavaType.java:24]: No, I think that was a typo. Fixed.

File: java/src/main/native/ai_onnxruntime_OnnxTensor.c (hunk 1 above)
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OnnxTensor.c:307]: Fixed.

File: java/src/test/java/ai/onnxruntime/ModelGenerators.java (hunk 2 above)
Comment (Craigacp) [reply java/src/test/java/ai/onnxruntime/ModelGenerators.java]: Thanks

File: java/src/main/java/ai/onnxruntime/OnnxJavaType.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply java/src/main/java/ai/onnxruntime/OnnxJavaType.java]: Can this be automatically inferred by the compiler?

File: java/src/main/java/ai/onnxruntime/OrtUtil.java [hunk 4]
```diff
@@ -15,9 +19,45 @@
...
+    MethodHandles.Lookup lookup = MethodHandles.lookup();
+    try {
+      // Attempt to lookup the Java 20 fp16 conversion methods which can use SIMD intrinsics.
+      tmp16 =
+          lookup.findStatic(
+              Float.class, "float16ToFloat", MethodType.methodType(float.class, short.class));
+      tmp32 =
+          lookup.findStatic(
+              Float.class, "floatToFloat16", MethodType.methodType(short.class, float.class));
+    } catch (IllegalAccessException | NoSuchMethodException e) {
```
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OrtUtil.java:44]: >IllegalAccessException | NoSuchMethodException e [](http://example.com/codeflow?start=13&length=48)

This is groovy!

File: java/src/main/java/ai/onnxruntime/OnnxJavaType.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxJavaType.java]: I rearranged the static init block so it constructs the array based on `OnnxJavaType.values().length`.

[the remark this answers is not in the document]
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxJavaType.java]: I also added a little bit of documentation to the enum types to fit in with the other docs PR I just created - #16776.

**Review by skottmckay — APPROVED** [review]

**Comment by skottmckay** [comment]
Re-opening. Trying to work around azp run issue.

**Comment by YUNQIUGUO** [comment]
Hi looks like when checking-in, we didn't run the checks for React Native CI pipeline.

It breaks at building AAR stage now, See: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1071335&view=logs&j=06bda0f0-c359-5ee8-d76d-283c2733373e&t=af31bba3-008c-59d3-c2d1-f81f08a354ff&l=6577

also expects to break the Zip-Nuget-Java-Nodejs Packaging Pipeline soon.

I am not too sure how to fix it. Any ideas? @Craigacp Thanks!

**Comment by Craigacp** [comment]
Looks like the minimum Android version it's set to doesn't support all of Java 7 which came out in 2011. I wasn't expecting Android to be that far behind, so I'll have to think about it. What version of Android is the minimum supported in ORT?

**Comment by YUNQIUGUO** [comment]
thanks for the reply.

looks like the android_sdk_version specified in building the aar package is 21.
https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/android/default_full_aar_build_settings.json#L8

**Comment by Craigacp** [comment]
The `MethodHandle.invokeExact` call which the build fails on is supported from Android O (SDK 26). That's not integral to the PR, we can drop the method handles and live with the fact that it'll be slower than necessary on Java 20 or newer. However longer term requiring a baseline Android version from 2014 is pretty restrictive.

**Comment by edgchen1** [comment]
> yeah.., looks like the `android_min_sdk_version` specified in building the aar package is 21. https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/android/default_full_aar_build_settings.json#L8
> 
> maybe @edgchen1 do you know elsewhere we specify for minimum Android version? Or is it just 21.

Yes, that is where it is specified for the onnxruntime-android package. onnxruntime-mobile has the same version.

> However longer term requiring a baseline Android version from 2014 is pretty restrictive.

We can revisit the minimum Android version that we want ORT to support. FYI @skottmckay @natke

**Comment by skottmckay** [comment]
If [this](https://www.appbrain.com/stats/top-android-sdk-versions) is to be believed ~7% of Android devices are running something older than 8.0.

Is there a build param that could be used to enable this if it's an Android build and the min SDK is >= 26?

**Comment by Craigacp** [comment]
The `MethodHandle` codepath swaps in a method added in Java 20 to improve the speed. I doubt Android is going to get methods from Java 20 any time soon so the performance benefit is irrelevant for that platform. I'll look on Monday at refactoring it so the conversions live in a separate file then do some juggling in gradle so Android always compiles the version without `MethodHandle`.

**Comment by Craigacp** [comment]
I've worked up what should be a fix in #16832, but it might need a bit of iteration on the android build system changes as I'm unfamiliar with it.

**Pull request merged**